### DISCUSSION TITLE
getServerSession error on getInitialProps

### DISCUSSION THREAD

**Discussion by rizchaerul** [nextauthjs/next-auth#4068] [Help] [answered]
### Environment

```
System:
  OS: Windows 10 10.0.22000
  CPU: (8) x64 Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
  Memory: 2.12 GB / 7.86 GB
Binaries:
  Node: 14.17.1 - C:\Program Files\nodejs\node.EXE
  Yarn: 1.22.17 - ~\AppData\Roaming\npm\yarn.CMD
  npm: 6.14.13 - C:\Program Files\nodejs\npm.CMD
Browsers:
  Edge: Spartan (44.22000.120.0), Chromium (98.0.1108.56)
  Internet Explorer: 11.0.22000.120
npmPackages:
  next: ^11.0.0 => 11.1.4
  next-auth: ^4.0.1 => 4.2.1
  react: ^17.0.2 => 17.0.2
```

### Reproduction URL

https://github.com/rizchaerul/get-server-session

### Describe the issue

I tried `getServerSession` in `getInitialProps` to make session available on all pages in SSR, but I got error `Can't resolve 'v8'`, but it's fine if tried it in `getServerSideProps`. So is it a bug or is it just not possible to call `getServerSession` on `getServerSideProps`?

here's the code:

```
MyApp.getInitialProps = async (appContext: AppContext): Promise<any> => {
  const appProps = await App.getInitialProps(appContext)

  // Not error.
  const session = await getSession(appContext.ctx)
  // Error: Can't resolve 'v8'.
  const serverSession = await getServerSession(appContext.ctx, nextAuthOptions)

  return {
    ...appProps,
    pageProps: {
      ...appProps.pageProps,
      session,
    },
  }
}
```

### How to reproduce

Just run `npm install` and `npm run dev` to get the error.

### Expected behavior

I expect `getServerSession` to work on `getInitialProps`, but if it's not possible is there work around this so I don't have to use `getInitialProps` on every pages?

**Comment by balazsorban44** [comment, marked as answer]
getInitialProps also runs in the browser, so it's not possible, no.

You can rather try our Middleware instead to lock down the entire app in a single line

`pages/_middleware.js`
```js
export { default } from "next-auth/middleware"
```

**Reply by rizchaerul** [reply]
I see, thank you!

**Comment by ellisio** [comment]
@balazsorban44 Since Next JS v12 removed `_middleware.ts` in favor of the Edge `middleware.ts`, how can we do this in `_app.tsx::getInitialProps()` if we're using a database adapter?

Doing it on every page is nice, but that is a lot of copy and paste and it would be cleaner to centralize this logic.
